### PR TITLE
add contexts for weather information

### LINKED ISSUE #82: Adding a Weather contexts file
Adding a Weather contexts file, containing (obviously) all possible contexts for the weather domain.

### PR DESCRIPTION
**Pull request by ddumet**
Adding a weather context, specific to weather attributes/vocabularies, that can (should!) be reused in any project/domain that needs it.
It is made up of a specific weather.jsonld contexts file that contains terms that are not already present into contexts `https://fiware.github.io/data-models/context.jsonld` and `https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context.jsonld`, and a weather-compound.jsonld contexts file, which is the one to use.

resolve #82

### CODE REVIEW

File: weather/jsonld-contexts/weather-compound.jsonld [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "@context": [
+        "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/weather/jsonld-contexts/weather.jsonld",
+        "https://fiware.github.io/data-models/context.jsonld",
```
Comment (bobeal) [inline weather/jsonld-contexts/weather-compound.jsonld:4]: did you have a look at the newer data models in the smart data models initiative?

this one still exists but is planned to disappear at some point.

Comment (bobeal) [reply weather/jsonld-contexts/weather-compound.jsonld:4]: https://github.com/smart-data-models/dataModel.Weather

Comment (bobeal) [reply weather/jsonld-contexts/weather-compound.jsonld:4]: and the context is here: https://github.com/smart-data-models/dataModel.Weather/blob/master/context.jsonld

Comment (bobeal) [reply weather/jsonld-contexts/weather-compound.jsonld:4]: we can also contribute the missing terms, it avoids maintaining a specific subset

File: weather/ngsild-payloads/weatherObserved-multiple-sources [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    "datasetId": "urn:ngsi-ld:Dataset:WeatherStation",
+    "observedBy": {
+      "type": "Relationship",
+      "object": "urn:ngsi-ld:Device:WeatherStation:dc94e9f5-6e3a-443d-88b4-778d2519b3c6"
+    },
+    "value": 12,
+    "observedAt": "2022-01-05T23:00:00Z",
+    "unitCode": "E92"
+  },
+  "relativeHumidity80": {
```
Comment (bobeal) [inline weather/ngsild-payloads/weatherObserved-multiple-sources:15]: what is the difference between `relativeHumidity80` and `relativeHumidity90`? could not it be the same property (with 2 instances)?

File: weather/ngsild-payloads/weatherObserved-multiple-sources [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    "datasetId": "urn:ngsi-ld:Dataset:WeatherStation",
+    "observedBy": {
+      "type": "Relationship",
+      "object": "urn:ngsi-ld:Device:WeatherStation:dc94e9f5-6e3a-443d-88b4-778d2519b3c6"
+    },
+    "value": 144,
+    "observedAt": "2022-01-05T23:00:00Z",
+    "unitCode": "SEC"
+  },
+  "windAverage1mn": {
```
Comment (bobeal) [inline weather/ngsild-payloads/weatherObserved-multiple-sources:37]: a bit weird to have the `1mn` in the property name. The property name could be `windAverage` and `1mn` a sub-property of the property.

File: weather/jsonld-contexts/weather-compound.jsonld (hunk 1 above)
Comment (ddumet) [reply weather/jsonld-contexts/weather-compound.jsonld:4]: Yes I did. My understanding was that the https://fiware.github.io/data-models/context.jsonld context contains all terms from fiware, i.e. covering the ones defined in https://github.com/smart-data-models/dataModel.Weather/blob/master/context.jsonld.
So, if the former will disappear, we will have to include several fiware contexts, right ? E.g. the ones for weather, agrifood, etc.

File: weather/ngsild-payloads/weatherObserved-multiple-sources (hunk 2 above)
Comment (ddumet) [reply weather/ngsild-payloads/weatherObserved-multiple-sources:15]: Could be, I just went for the Ockham's razor principle. I didn't want to over complicate the model with property/ies of property.
Same for windAverage1m.
The thing with models is that there are several solutions, which one is best is quite subjective. E.g. There could be a property Wind, with sub-property speed, which itself has sub-properties instantaneous and average. Another direct property of wind would be direction.
Is this level of complexity worth it ?

I'll come up with another more complex model next week, then we can compare and decide what's best then.

Comment (bobeal) [reply weather/ngsild-payloads/weatherObserved-multiple-sources:15]: > Could be, I just went for the Ockham's razor principle. I didn't want to over complicate the model with property/ies of property. Same for windAverage1m. The thing with models is that there are several solutions, which one is best is quite subjective. E.g. There could be a property Wind, with sub-property speed, which itself has sub-properties instantaneous and average. Another direct property of wind would be direction. Is this level of complexity worth it ?
> I'll come up with another more complex model next week, then we can compare and decide what's best then.

The point is not about the complexity, but about a clear semantic and extensibility.

I don't really want a data model with dozens versions of `relativeHumidityXX` or as many `windAverageX` as they are possible duration in the universe.

File: weather/jsonld-contexts/weather-compound.jsonld (hunk 1 above)
Comment (bobeal) [reply weather/jsonld-contexts/weather-compound.jsonld:4]: The one on fiware.github.io is no longer maintained (`Last-Modified: Mon, 12 Apr 2021 12:38:25 GMT`), contexts are now under the Smart Data Models umbrella.

So, yes, we have to include the subdomains we need in our compound contexts (e.g., https://github.com/smart-data-models/dataModel.Weather/blob/master/context.jsonld).

File: weather/jsonld-contexts/weather.jsonld [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+    "@context": {
+        "observedBy": "https://uri.fiware.org/ns/data-models#observedBy",
+        "cloud": "https://uri.fiware.org/ns/data-models#cloud",
+        "visibility": "https://uri.fiware.org/ns/data-models#visibility",
+        "precipitationIntensity": "https://uri.fiware.org/ns/data-models#precipitationIntensity",
+        "relativeHumidity90": "https://uri.fiware.org/ns/data-models#relativeHumidity90",
```
Comment (bobeal) [inline weather/jsonld-contexts/weather.jsonld:7]: don't forget to update this

Comment (ddumet) [reply weather/jsonld-contexts/weather.jsonld:7]: Yes, wanted to make sure we agree on the model by itself before updating the context file(s).
@AHABID, are you happy with this model as well ?

Comment (AHABID) [reply weather/jsonld-contexts/weather.jsonld:7]: Yes, the model seems to be greate

Comment (ddumet) [reply weather/jsonld-contexts/weather.jsonld]: I've updated Contexts.
Can we still use "observedAt" and "observedBy" ?
I noticed in the smartData.Weather context, there's a "dataProvider" and "dateObserved" ...

Comment (bobeal) [reply weather/jsonld-contexts/weather.jsonld]: `observedAt` is part of the NGSI-LD specification, so yes we can, and even must.

`dataProvider` could be a replacement for `observedBy`, but as we widely use `observedBy`, I suggest to stick to it.

**Comment by bobeal** [comment]
LGTM.

**Review by bobeal — APPROVED** [review]

**Review by AHABID — APPROVED** [review]

**Pull request merged**